WalletWasabi/WabiSabi/Backend/DoSPrevention/Warden.cs

```diff
@@ -2,42 +2,53 @@
 using System.IO;
 using System.Linq;
 using System.Threading;
+using System.Threading.Channels;
 using System.Threading.Tasks;
-using WalletWasabi.Bases;
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
 using WalletWasabi.Helpers;
 using WalletWasabi.Logging;
+using WalletWasabi.WabiSabi.Backend.Rounds.CoinJoinStorage;
 
 namespace WalletWasabi.WabiSabi.Backend.DoSPrevention;
 
-/// <summary>
-/// Serializes and releases the prison population periodically.
-/// </summary>
-public class Warden : PeriodicRunner
+public class Warden : BackgroundService
 {
-	/// <param name="period">How often to serialize and release inmates.</param>
-	public Warden(TimeSpan period, string prisonFilePath, WabiSabiConfig config) : base(period)
+	public Warden(string prisonFilePath, ICoinJoinIdStore coinjoinIdStore, WabiSabiConfig config)
 	{
 		PrisonFilePath = prisonFilePath;
-		Config = config;
-		Prison = DeserializePrison(PrisonFilePath);
-		LastKnownChange = Prison.ChangeId;
+		OffendersToSaveChannel = Channel.CreateUnbounded<Offender>();
```

Comment:
Ok. I will do it but it is hard to see how that makes sense because it is the **producer** the one that should call `Complete` to let the consumer to know that there will not be more items but in this case it is the **consumer** who tells itself that the producer will not add more items. It's weird.